### PR TITLE
[mcs] Fixes codegen for pattern probing with value-type variables

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

File: mcs/mcs/expression.cs [hunk 1]
```diff
@@ -1692,37 +1692,70 @@ void EmitLoad (EmitContext ec)
...
+						Variable.CreateBuilder (ec);
 
-				// Only to make verifier happy
-				if (probe_type_expr.IsGenericParameter && TypeSpec.IsValueType (expr.Type))
-					ec.Emit (OpCodes.Box, expr.Type);
+					Variable.EmitAssign (ec);
+				}
 
-				ec.Emit (OpCodes.Isinst, probe_type_expr);
+				ec.MarkLabel (no_value_label);
```
Comment (lewurm) [inline mcs/mcs/expression.cs]: not visible in the diff, but you can move the `no_value_label` declaration into the `expr_unwrap` case and also get rid of a `new Label ()` allocation.

File: mcs/mcs/expression.cs [hunk 2]
```diff
@@ -1733,11 +1766,20 @@ void EmitLoad (EmitContext ec)
...
+					var end = ec.DefineLabel ();
+
+					ec.EmitInt (1);
+					ec.Emit (OpCodes.Br_S, end);
+					ec.MarkLabel (nonmatching_label);
+					ec.EmitInt (0);
+					ec.MarkLabel (end);
+				} else {
+					ec.MarkLabel (nonmatching_label);
 				}
```
Comment (lewurm) [inline mcs/mcs/expression.cs]: I don't understand why this special casing for `vtype_variable` is needed. can't we always emit `dup` and bail out to the `nonmatching_label` at the very end?

File: mcs/mcs/expression.cs (hunk 1 above)
Comment (marek-safar) [reply mcs/mcs/expression.cs]: yeah, it's struct but it looks better

File: mcs/mcs/expression.cs (hunk 2 above)
Comment (marek-safar) [reply mcs/mcs/expression.cs]: good find, it's probably slower but not that much

**Comment by marek-safar** [comment]
@monojenkins build Linux x64 mcs

**Pull request merged**